[IMP] website, website_event: add rel=canonical tag

seb-odoo opened the pull request:
The canonical tag is important for SEO, indeed it prevents search engines from
indexing duplicate content.

Reasoning
=========

The choice has been made to create the canonical tag automatically depending on
the request path, ignoring the query string, and manually prefixing the
appropriate domain and language code.

Indeed creating it manually for each resource would create a lot of code and
potential mistakes.

It is more dangerous to do it the generic way, but after investigation it
appears that it is an acceptable trade-off since the vast majority of our routes
are well built and already ready for this:

- using query string only for minor features that do not change the main content
- having the models, the ids, the pager and other important features in the path

Override
========

It is still possible to override the default behavior by passing
`canonical_params` manually to the view or to the different methods.

This is done for `/event` because the only way to display Past Events is to add
`date=old`.

Languages
=========

Fix an issue where it was possible for a bot to be on the URL without language
code but to use a language that is not the default language.

Adapt hreflang, because it:
- must only be present on canonical pages
- must always lead to canonical pages
- should not be set if there is no alternate language

Misc
====

task-1958075
closes #12532

Inspired by OCA module `website_canonical_url` courtesy of Jairo Llopis.

Co-authored-by: Jairo Llopis <jairo.llopis@tecnativa.com>
Co-authored-by: Sébastien Theys <seb@odoo.com>

pedrobaeza commented:
Great work! I let technical review to my colleague @Yajo

Yajo commented:
I'll appreciate if you can remove my mention (@Yajo) from the commit. Just put Jairo Llopis please. Github is spamming me on each clone/rebase/push/amend/etc. of that commit. :sweat_smile:

seb-odoo commented:
Done, sorry for the spam. Hopefully it is better now.

Yenthe666 commented:
Sweet!

JKE-be commented on addons/website/tests/test_base_url.py:
```diff
@@ -1,22 +1,26 @@
+    def setUp(self):
+        super(TestBaseUrl, self).setUp()
+        self.domain = 'https://www.domain.jke'
+        self.website = self.env['website'].browse(1)
```
keep  Website.create({'name': 'test base url', 'domain': domain})

Borwse Id instead of search([], limit=1) is a bad idea
And avoid to depend of data inyour test

Yajo replied:
You might also use:
```suggestion
        self.webstie = self.env.ref("website.default_website")
```

seb-odoo replied:
I'll keep the code to create a new website but set the domain to `'http://' + odoo.tests.HOST` so it is actually selected as the current website when doing `url_open`, which is what I wanted by using `browse(1)`.

JKE-be commented:
@seb-odoo 

Part of the task was also to check the question about duplicate content in case of ?sort=latest_blog page for example. Did you check it ?

seb-odoo commented:
For lang, I'll check today, for runbot hopefully my last force push will fix it

seb-odoo commented:
All references I checked agree that listing pages (also called search pages) should be canonical per page but not per filter or sort.

The goal is only one listing indexed (no duplicate) but all its page are indexed (full content coverage).

Some references:
- https://moz.com/blog/which-page-is-canonical
- https://support.google.com/webmasters/answer/139066?hl=en
- https://tools.ietf.org/html/rfc6596

This is how the current implementation is going to work if we usually filter or sort by query string and have the pager in the path.

In the specific context of our /blog route, it could be argued that the route by /tag should in fact not be canonical because it is duplicate content from the main list.

Same could be said for the multi blog route and specific blog route when multiple blogs are used.

Yajo commented:
Sorry to get a little bit off-topic :innocent:. Let me gently abuse this PR to spam you about https://github.com/odoo/odoo/pull/30832, which is another big "crawlability" problem in Odoo. Going back to my cave. :monkey:

seb-odoo commented:
Pushed commit to improve the situation between hreflang and canonical, but it is impossible to make hreflang work correctly with our existing choices.

To follow standards we need to either:
- use the URL without language code for the default lang, and remove the magic redirect to user language
- use the URL without language code to do the magic redirect (and define it as hreflang x-default), and have a proper URL with language code for the default language

Currently it doesn't make sense to say "this URL is for this language" and then have the exact same URL redirect the user to another language automatically.

Yajo commented on addons/website/models/website.py:
```diff
@@ -445,11 +445,23 @@ def get_languages(self):
+            return []
+
+        languages = self.get_languages()
+        if len(languages) <= 1:
```
Actually `if self.language_ids:`

JKE-be replied:
<= 1

seb-odoo replied:
Ok, I need to rebase on https://github.com/odoo/odoo/commit/269aa594111a152ad4b7714856ea745bfef57155

seb-odoo commented:
After even more investigation I think we should just remove the magic redirect.

And if for some reason we must absolutely keep it, no hreflang should point to the magic URL without language code (as opposed to what I did in this PR), and that URL should only redirect and not sometimes display content. That implies the default lang should have an URL with the language code in it.

Some ref: https://support.google.com/webmasters/answer/182192?hl=en&ref_topic=2370587

> **Use different URLs for different language versions**

> Google recommends using different URLs for each language version of a page rather than using cookies or browser settings to adjust the content language on the page.

> If you prefer to dynamically change content or reroute the user based on language settings, **be aware that Google might not find and crawl all your variations**.

cc @rdeodoo I think you should follow this PR

rdeodoo commented:
> Google recommends using different URLs for each language version of a page

That's what we do, isn't it? /fr/my-super-blog-1 and /en/my-super-blog-1, completely different URL?

I am not against having the main language displayed too, like alternate language (maybe unless there is only 1 lang ofc). Most of the website are doing that anyway, every lang will be displayed, even the main one AFAIR.

About the magic redirect, it only happens if the URL has no lang in it. The main language has like a 'pass-droit' which will redirect to the user lang.

**If the main lang was displayed in the URL, wouldn't that solve the magic redirect issue of your crawler? Eg it should not land on URL without lang.**

I think the magic redirect is cool if you want to send a mass mailing, you just use the straight url with no lang and every user will type it and still be on its selected language.

It really annoys you regarding the crawler, isn't it?

Yajo commented:
I also think that the magic redirect is great, yes. This way, you can always link to the lang-agnostic page, and the user gets to the lang-specific one. That's great.

However I don't think it just works for a crawler, because that magic redirection is (and must be) a 302, so that doesn't indicate to the crawler that the lang-agnostic mustn't be indexed.

So, the roadmap would be:

1. Add a noindex instruction to lang-agnostic pages.
1. Always add the lang prefix, no matter if it's the default or if website has only 1 lang.

That should be enough for all cases, isn't it?

rdeodoo commented:
> * Add a noindex instruction to lang-agnostic pages.
> * Always add the lang prefix, no matter if it's the default or if website has only 1 lang.

I think that would be the correct way yes, except for only 1 lang, I don't think it should display the URL lang.

Yajo commented:
But if you only have 1 lang, it gets indexed, and later you add a new lang, then the crawler would get confused. That's why I think it's better to keep it since the beginning.

JKE-be commented:
> Add a noindex instruction to lang-agnostic pages.

no, with an hreflang x-default, google understand that the current page can sometime redirect to an appropriate language, sometimes not...

So for me we need to continue to index it.


The only problem is for the hreflang='en' (if en default lang), but it is already the case at the moment... 

So we will keep it like that, or just remove this one the hrelang='<default-lang>'... 

> Always add the lang prefix, no matter if it's the default or if website has only 1 lang.

problem is that a canonical url shouldn't have a redirect what will be the case if you point to the /en url, so we keep canonical without en

Yajo commented:
True, I forgot about x-default... So a new roadmap:

1. Add `hreflang="x-default"` to lang-agnostic pages.
2. Always add the lang prefix, no matter if it's the default or if website has only 1 lang.

If we do it like this, what would be the problem with the canonical URL? I think I don't understand that part. :thinking:

JKE-be commented:
That actually we don't want the /<default-lang> in url

that's why we have the is_a_bot function

Yajo commented:
Which /? Sorry I don't understand you, could you explain yourself a little bit? :roll_eyes: :point_right: :point_left:

JKE-be commented:
we don't want to have an url www.odoo.com/en when you have only one lang (en even when you have several lang)

And we don't want to have a website www.odoo.com while you only have on lange, and when you will install a second lang, all existing url should be redirected from odoo.com to odoo.com/en

So the idea, is just to keep the actual system domain.com/ = default lang; and use hacky function is_a_bot (not recommanded by google) to make other behavior for search engine and always show the odoo.com/ in english withtout try to redirect... (hack introduce in the same time of this redirect feature to avaoid this problem)

Yajo commented:
OK I think I understand now... However that has a problem too. Many links across odoo are rendered without lang prefix, relying in the magic redirection (and that's how it should be IMHO). But if you don't redirect the bot, then some lang-specific pages would never get indexed, right? Or am I missing something?

JKE-be commented:
>  then some lang-specific pages would never get indexed

That's why we have href-lang -> to show to google that the page have a translated page in other language ;)

seb-odoo commented:
Force pushed. Everything should be working fine now thanks to `is_a_bot`, but code had to be changed to make sure when a bot is not redirected that it uses the default language (that patch should be _considered_ for backport/merged in earlier version because there could be, in theory, indexing issue with the default language before that fix).

There is only one thing that is not really working on the PR is when the URL path contains characters that can be quoted (such as `,`) but this is only an issue for the blog tags and those have been removed in the proposed PR above (v10, waiting forward port).

So I think (and hope to not find anything else) that this PR is finally in an acceptable functional state.

JKE-be commented on addons/website/models/website.py:
```diff
@@ -437,37 +438,53 @@ def page_search_key_dependencies(self, page_id=False):
+        self.ensure_one()
+
+        # Compare query string dict to make sure order is not taken into account.
+        params = urls.url_decode(request.httprequest.query_string)
```
params = request.params
canonical_params = canonical_params or {}

or

params = request.args
canonical_params = canonical_params or werkzeug.datastructures.MultiDict()

JKE-be commented on addons/website/models/website.py:
```diff
@@ -437,37 +438,53 @@ def page_search_key_dependencies(self, page_id=False):
+        params = urls.url_decode(request.httprequest.query_string)
+        canonical_params = canonical_params or werkzeug.datastructures.MultiDict()
+        # Compare URL has to be done without lang in URL because the lang code
+        # has been removed from request.httprequest.base_url during dispatch.
```
bof :/ can't we use e.g. 

current_url = request.rerouting and request.rerouting[0] or request.httprequest.path

seb-odoo replied:
As discussed, used `root_url` + rerouting stuff to get rid of `lang_in_url`

JKE-be commented on addons/website/models/website.py:
```diff
@@ -827,6 +844,36 @@ def get_base_url(self):
+            arguments = dict(request.endpoint_arguments)
+            for key, val in list(arguments.items()):
+                if isinstance(val, models.BaseModel):
+                    arguments[key] = val.with_context(lang=lang.url_code)
```
check that not already the same lang in actual context ?

JKE-be commented on addons/website/models/website.py:
```diff
@@ -827,6 +844,36 @@ def get_base_url(self):
+            path = urls.url_quote_plus(request.httprequest.path, safe='/')
+        lang_path = ('/' + lang.url_code) if lang_in_url else ''
+        canonical_query_string = '?%s' % urls.url_encode(canonical_params) if canonical_params else ''
+        return self.get_base_url() + lang_path + path + canonical_query_string
```
as discuss previously, can we split canonical job, and absolute domain... 
E.g. 
`_get_canonical_url`  that `return lang_path + path + canonical_query_string`
and 
`_get_absolute_canoncial_url` that `return  self.get_base_url() + _get_canonical_url`

JKE-be commented:
for i in range(0, 1000):
    robodoo r+

JKE-be commented:
@robodoo please r+ and retry this branch

robodoo commented:
I'm sorry, @JKE-be. Retry makes no sense when the PR is not in error.

robodoo commented:
'ci/runbot' failed on this reviewed PR.

robodoo commented:
'ci/runbot' failed on this reviewed PR.

robodoo commented:
'ci/runbot' failed on this reviewed PR.

robodoo commented:
'ci/runbot' failed on this reviewed PR.

JKE-be commented:
@robodoo r+ retry

robodoo commented:
I'm sorry, @JKE-be. Retry makes no sense when the PR is not in error.

robodoo commented:
'ci/runbot' failed on this reviewed PR.

robodoo commented:
'ci/runbot' failed on this reviewed PR.

seb-odoo commented:
We have CI!

JKE-be commented:
@robodoo r+ and give you a beer

robodoo commented:
Merged at 1781041f13b66dfe91b3230794ffb6b2001d86c1, thanks!

Yenthe666 commented:
Now you own him a beer @JKE-be! Perhaps at Odoo Experience 😉

seb-odoo commented:
@zeroheure Do you have a direct link to an article on that website (or any other) about the sorting and filtering problem, and specifically how/why canonical is not solving it?

Also we've not implemented exactly like the OCA module, so the result might not be the same.

zeroheure commented:
@seb-odoo There is not much freely available on abondance.com 
Here's a [recent video](https://www.abondance.com/20190507-39641-seo-et-e-commerce-2e-partie-la-liste-de-produits-ideale-video-seo-numero-123.html) (french), Olivier Andrieu talks about "navigation à facette" (sort and filtering) at 4:08. It is basic but expose the problem. (I have also emails and a website audit from him).
There is a lot more [on the private part of his site](https://www.reacteur.com/?s=facette).

Here's recent [Google advices about faceted navigation](https://webmasters.googleblog.com/2014/02/faceted-navigation-best-and-5-of-worst.html), where it is clear that canonical _alone_ doesn't solve the problem.

seb-odoo commented:
After reading the last article, it seems that, apart from `rel="canonical"`, the two other main tips are:
- avoid generating links that lead to empty pages, such as done in https://github.com/odoo/odoo/pull/36561 for categories
- use `rel="nofollow"` on all the filtering links (and/or `Robots.txt` exclude them if possible).

We usually try to do those when appropriate, though there might still be room for improvement.

They also suggest to use parameters instead of using the path for categories and other meaningful pages: apparently they can guess more easily what to crawl then, but this is at the opposite of what we do in our routes.
But more importantly they also recommend to have meaningless parameters not in the path, which is what we do. Also more important parameters should be listed first.

Finally, for pagination, it is recommended to add canonical on all pages to a "view all" page, but only if it is doable, without impacting performance. So this cannot apply in our case, unless we add more complex logic based on the number of products/events/posts/... and only canonicalize pages if less than an arbitrary number so that the "view all" is not slow.

zeroheure commented:
Thankyou for taking time to read it and clarifying problem. So your current implementation is the best that can be done ? And the only remaining problem is that we may have to add parameters in Google Search Console.

seb-odoo commented:
I would not say it's the best that can be done, but we are going in the right direction for sure.

There are probably other links where we could add a "nofollow", but we will not have time to do that before version 13.
It is safer to not use them and use a bit more crawler time on duplicate links (that are now canonicalized!), instead of risking to break indexation of some pages because they can't be reached anymore.

The current compromise seems good enough, we have fixed the worse occurrences of exponentially growing links (such as in #30832 referenced above for blog, and in the current PR for slides), and added canonical to send the correct info for the remaining duplicate pages.

As said, we could go further for search pagination by being a bit smarter when there are few items listed, but this would be at the cost of more complexity and potential mistakes.

Since we ignore all parameters by default in our implementation of canonical, I don't think there is anything to do in the search console regarding that.

Feedback is always welcome post v13 to keep improving it where needed.

Yajo commented:
There's still a big problem then: `website_sale` attribute filters:

![Captura de pantalla de 2019-10-04 09-36-00](https://user-images.githubusercontent.com/973709/66193430-6a9c1500-e68a-11e9-8cba-0a054aeb56b2.png)

They are no `<a>`, they are `<input>` or similar elements. How are we supposed to add a `rel="nofollow"` to them? Does it work in inputs as well?

The canonical improvement would still save us from that problem in many cases, but the nofollow there would save a lot of time that crawlers spend combining filters.

zeroheure commented:
Google only follows links if there is an href, using form elements is enough to stop the search engine bot (another way is to use JS with `document.location`).

Yajo commented:
OK the logs I saw were not attrs indeed, they were the order. I opened #37984 for that case.
